server/aesdsocket.c

```diff
@@ -22,51 +27,240 @@
+	}
+
+	// Write packet to file
+	pthread_mutex_lock(&g_sockdata_mutex);
+	g_data_file = fopen(SOCK_DATA_FILE, "a+");
+	if(g_data_file == NULL) {
+		fprintf(stderr, "Couldn't open file\n");
+		return thread_param;
+	}
+	fwrite(dyn_recv_buf, 1, recv_size_tot, g_data_file);
+	free(dyn_recv_buf);
+	dyn_recv_buf = NULL;
+	fclose(g_data_file);
+	g_data_file = NULL;
+	pthread_mutex_unlock(&g_sockdata_mutex);
+
+	// Send back data from file
+	// Note: line-by-line read code learned from https://stackoverflow.com/questions/3501338/c-read-file-line-by-line
+	char *line = NULL;
+	size_t len = 0;
+	ssize_t read;
+	pthread_mutex_lock(&g_sockdata_mutex);
+	g_data_file = fopen(SOCK_DATA_FILE, "r");
+	if(g_data_file == NULL) {
+		fprintf(stderr, "Couldn't open file\n");
+		return thread_param;
+	}
+	while((read = getline(&line, &len, g_data_file)) != -1) {
+		if(send(tdata_ptr->accepted_sock_fd, line, read, 0) == SOCK_ERR) {
+			fprintf(stderr, "socket send err\n");
+		}
+	}
+
+	fclose(g_data_file);
+	pthread_mutex_unlock(&g_sockdata_mutex);
+	g_data_file = NULL;
+	free(line);
+
+
+	tdata_ptr->complete = true;
+	close(tdata_ptr->accepted_sock_fd);
+
+	// Log connection close
+	syslog(LOG_DEBUG, "Closed connection from %s\n", tdata_ptr->ip_addr);
+
+	return thread_param;
+}
 
 // Signal handler for SIGINT and SIGTERM
 // Free all dynamic memory, close all files & sockets
 void handle_kill(int sig) {
+	struct tdata *tdata_cur = NULL;
+	void *t_ret = NULL;
+
 	syslog(LOG_DEBUG, "Caught signal, exiting\n");
-	if(g_accepted_sock_fd)
-		close(g_accepted_sock_fd);
+
+	while (!SLIST_EMPTY(&g_head)) {
+        tdata_cur = SLIST_FIRST(&g_head);
+        if(pthread_join(tdata_cur->thread, &t_ret) != 0) {
+			fprintf(stderr, "thread join fail. cant free memory without thread. bad!\n");
+		} else {
+	        SLIST_REMOVE(&g_head, tdata_cur, tdata, ptrs);
+	        destroy_tdata(tdata_cur);
+    	}
+    }
+
 	if(g_sockfd)
 		close(g_sockfd);
 	if(g_data_file != NULL)
 		fclose(g_data_file);
 	if(g_servinfo != NULL)
 		freeaddrinfo(g_servinfo);
-	if(g_dyn_recv_buf != NULL)
-		free(g_dyn_recv_buf);
-
+	pthread_mutex_lock(&g_sockdata_mutex);
 	if(remove(SOCK_DATA_FILE) != 0) {
 		fprintf(stderr, "Error deleting %s\n", SOCK_DATA_FILE);
 		exit(1);
 	}
+	pthread_mutex_unlock(&g_sockdata_mutex);
+	pthread_mutex_destroy(&g_sockdata_mutex);
+
+
+	pthread_cancel(g_timer_thread);
 
 	exit(0);
 }
 
-// Check if string contains char c
-// return 1 if it does, otherwise 0
-int contains_chr(char *str, char c, int len) {
-	for(int i=0; i<len; i++) {
-		if(str[i] == c) {
-			return 1;
+// Thread for timer
+void *timer_thread(void *data) {
+	time_t t;
+	struct tm *tm_p;
+	char time_str[30] = {0};
+	struct tdata *tdata_cur = NULL;
+	struct tdata **tdata_ptrs = NULL;
+	void *t_ret = NULL;
+	int num_join = 0;
+
+	while(1) {
+		sleep(10);
+
+		time(&t);
+		tm_p = localtime(&t);
+		strftime(time_str, sizeof(time_str), "%a, %d %b %Y %T %z", tm_p);
+
+		pthread_mutex_lock(&g_sockdata_mutex);
```

Comment:
@betai97 what would happen if your `SIGTERM` signal occurred immediately after this line?